packages/loader/driver-utils/src/runWithRetry.ts

```diff
@@ -30,40 +30,77 @@ export interface IProgress {
     /**
      * Called whenever api returns cancellable error and the call is going to be retried.
      * Any exception thrown from this call back result in cancellation of operation
-     * and propagation of thrown exception.
      * @param delayInMs - delay before next retry. This value will depend on internal back-off logic,
      * as well as information provided by service (like 429 error asking to wait for some time before retry)
      * @param error - error object returned from the call.
      */
     onRetry?(delayInMs: number, error: any): void;
 }
 
-export async function runWithRetry<T>(
+export interface IProgress2 {
+    /**
+     * Abort signal used to cancel operation
+     * Note that most of the layers do not use this signal yet. We need to change that over time.
+     * Please consult with API documentation / implementation.
+     * Note that  number of layers may not check this signal while holding this request in a queue,
+     * so it may take a while it takes effect. This can be improved in the future.
+     * Layers in question are:
+     *    - driver (RateLimiter)
+     *    - runWithRetry
+     */
+    cancel?: AbortSignal;
+
+    /**
+     * Called whenever api returns cancellable error and the call is going to be retried.
+     * @param delayInMs - delay before next retry. This value will depend on internal back-off logic,
+     * as well as information provided by service (like 429 error asking to wait for some time before retry)
+     * @param error - error object returned from the call.
+     * @returns - true to continue retrying, false to abort.
+     */
+    onRetry?(delayInMs: number, error: unknown): boolean;
+}
+
+export type RunResult<T> = {
+    status: "succeeded";
+    result: T;
+} | {
+    status: "failed";
+    error: unknown;
+} | {
+    status: "aborted";
+};
+
+export async function runWithRetry2<T>(
     api: (cancel?: AbortSignal) => Promise<T>,
```

Comment:
since we give the api the abort signal, it should be able to return a RunResult as well.